### ISSUE TITLE
As a user, I want the members list page updated as soon as I change search terms, filters, or results page so that I can refine my search without losing my place in the form and have results load more quickly.

### ISSUE THREAD

**Issue by bwhicks** [Princeton-CDH/mep-django#140]
# testing notes
there are a number of things you can do to test this, but ultimately common sense/what you expect to happen is probably the most important:

- try the next/previous page buttons; watch the new results load
- try the page dropdown selector
- try clicking the "has card" filter
- try typing a query into the keyword search

all of these should trigger an immediate refresh of the results. basically, we want to make sure:

- the next/prev buttons are enabled/disabled when they should be
- the current page looks right
- the number of total results looks right
- results load when you expect them to load.

the blue loading bar on the page controls is new; would love comments on it and the feel of the search in general.

## dev notes
- this include replacing the total number of results indicator with a "loading..." text, a la PPA, so that the user knows a search is in progress
- @gissoo will revisit the loading indicator when working on search filter functionality if necessary
- this also includes writing a route that will serve JSON of the results so that the results can be updated via ajax.
- so far we have both a keyword query and the "has card" filter, so the form should be hooked up to both of them in addition to the page & sort controls.

**Comment by thatbudakguy** [comment]
work on this is ongoing since it's implicated in the PPA reactive search - how important is it to have it working in MEP this iteration? we can certainly build all of the search features that are included in this iteration without it, but we have to have it eventually.

**Comment by rlskoeser** [comment]
Increasing from 5 to 8pts.

**Comment by clmahoney** [comment]
1) page results set too high to be meaningful. For the vast majority of cases, there will never be more than 50 results for any keyword (example, we tested with "Smith" which yields 53 results). You should paginate by 25 instead. 
--> I realize we have the page results set so high because when there is no query, we have 6000+ results. What should we do??
2) Next/Previous buttons should not appear if there is only 1 page of results --> the next/prev buttons are NOT disabled 
3) when you press "Next" the display number disappears, and returns when you press "Previous"
4) sometimes entering numerical characters yields 0 results; other times it yields endless "results are loading." A alphanumerical combination ("2893749r") yields endless "results are loading."

OTHERWISE: 
- results load well 
- total number of results is right 
- "has card" feature narrows results to members with cards 
- I can't really see the blue load - everything loads very fast. So I'm not sure what do say.

**Comment by clmahoney** [comment]
(Thanks to MARY for helping me break things on the website; she rocks)

**Comment by clmahoney** [comment]
Tried this in Safari: 
- Blue loading bar is cool! I think it's smooth and nice. 
- Unfortunately nothing would actually load when I searched. It was endless "results are loading," even though I put in "Stein"

**Comment by clmahoney** [comment]
Tried in Firefox 
- Everything loading well here, with the same bugs as in Chrome

**Comment by thatbudakguy** [comment]
thanks for testing thoroughly! wanted to post (hah) a postmortem on 2 for dev team before i get to the rest of this because it's sort of interesting.

**expected behavior**: when there's only one page of results, both the previous and next page buttons should be disabled.

**actual behavior**: the previous button is disabled, but the next button isn't.

**cause**: this line uses `else if`, when it needs to be just `if`, since both conditions can happen simultaneously:
https://github.com/Princeton-CDH/mep-django/blob/805c7181679f176a7877bc3b9a5ff40a6799527e/srcmedia/ts/components/PageControls.ts#L99

why didn't the tests catch this, though? turns out the `UnhandledPromiseRejectionWarning` I've seen popping up in CI - which doesn't cause the tests to fail - is a clue. all of the tests for `update()` are making assertions inside the `then()` callback, which is fine..._if_ if the promise resolves. if it doesn't, jest has no way to know that the promise was rejected, because there's no `catch()` handler and the promise isn't returned.
https://github.com/Princeton-CDH/mep-django/blob/805c7181679f176a7877bc3b9a5ff40a6799527e/srcmedia/ts/components/PageControls.test.ts#L98

**solution**: make sure to `return` the promise in which the test is making assertions, so that the failure can be caught.
```js
    return pc.update([1, 10]).then(() => { //...
```

**Comment by thatbudakguy** [comment]
sadly, this has implications for almost all the tests in `MainMenu.test.ts` too, some of which are now failing when they weren't before... 😱 I think you really helped us out here @clmahoney

**Comment by rlskoeser** [comment]
@thatbudakguy thanks for catching the if/else logic on my change, can't believe I missed that (trusted the tests too much?).  I also saw the warnings but saw them before and after my changed so figured they were ignorable...  Glad you found out the problem with the tests now, though!

**Comment by thatbudakguy** [comment]
tests are now fixed in develop. regarding the other comments on this issue:
1 - not sure what we can do about this, since doing anything less than ~100 results per page makes the dropdown basically unusable when you're first coming to the search. open to ideas.
2 - addressed above; should be fixed now.
3 - i think this is just a symptom of 2, right? if you're prevented from clicking "next" to a nonexistent page it shouldn't happen anymore either
4 - could this be a solr thing? @rlskoeser any ideas?

will look into the safari behavior.

**Comment by rlskoeser** [comment]
**1**. Regarding the number of results per page: I agree that 100 results per page is too high to be really usable, but we're also concerned about the number of pages total. My preference is to wait until after the person merge work is done so we know how many people we have with no search terms or filters.
**4**. I can get this to happen in staging but I'm not sure what's causing the problem. At least some of the time an empty result does properly update the 'results loading' to '0 results', but I don't see any obvious differences in the headers. There's no solr error @thatbudakguy - you can confirm by reloading the search page that causes the problem. Any chance your logic for updating the page labels doesn't like the empty page labels header? Not sure how that would account for it working some of the time, but I did notice that when the 'results loading' never changes, the page labels dropdown is empty; when I get "0 results" the page labels dropdown is unchanged from the previous result set. (Not actually sure what it _ought_ to do in that case...)

**Comment by thatbudakguy** [comment]
yeah, there's more going on here than I thought.
- the logic for receiving the page labels and total pages is wrong; it should default to `'0'` and `''` instead of not updating when it doesn't receive those headers
- the logic for updating the page labels doesn't handle this well either; it keeps the old page labels around when you have zero results, which lets you request impossible pages (same issue as #284)

talked to @gissoo and we probably need to add some text to display in the page dropdown when there are no pages to select, she suggested "N/A" and disabling the select.

**Comment by thatbudakguy** [comment]
ok, I've implemented the `"N/A"` solution.

**Comment by clmahoney** [comment]
1 - I agree with Rebecca that we should wait until we merge. 
2- the next/previous are blocked out when I load results for "stein" and  it shows me all of them on the page so it's looking good 
3- solved because of 2, as Nick said
4- entering alpha-numerical things doesn't screw anything up, just shows 0 results, so this is great

NOW: I tried this in Safari and nothing will load for me!! I'm not sure if this is a symptom of something bigger or just a specific issue with my Safari. Like everything just endlessly loads with the blue bar moving across the screen. 

Firefox looks all good.

**Comment by thatbudakguy** [comment]
hey @clmahoney ,

glad the next/prev stuff is working! unfortunately i can't replicate the safari problems. can you post the:

- version of MacOS you are using, e.g. `10.14.4` (it's in  → About This Mac)
- version of safari you are using (Safari → About Safari)

maybe also try it on a different computer and/or an iPhone (via mobile safari) if you have access to one.

**Comment by rlskoeser** [comment]
@clmahoney one of us can look at your machine with you the next time you're here in person so we can try to figure out what's going on with Safari.

**Comment by clmahoney** [comment]
My MacOS is 10.11.6. My safari is  11.1.2. Thanks to the both of you @thatbudakguy and @rlskoeser for helping! 

I'll try a different computer too.

**Comment by clmahoney** [comment]
Signing off on this because I tried Safari on my iPad mini and it works great.
